### ISSUE TITLE
Rule Change: `no-unused-expressions ` to handle useless assignments

### ISSUE THREAD

**Issue by boris-petrov** [eslint/eslint#17539]
### What rule do you want to change?

no-unused-expressions

### What change to do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
function x(arg) {
  arg = 1;
  return 8;
}
```

### What does the rule currently do for this code?

Nothing.

### What will the rule do after it's changed?

Note that the assignment to `arg` (which is an argument/local variable) is useless - it's unused after that.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

I just found a bug in my code because of that. I found it strange that there isn't a lint rule for it. :)

**Comment by Rec0iL99** [comment]
Hi @boris-petrov, thanks for the issue. This looks like a scenario for the [`no-unused-vars`](https://eslint.org/docs/latest/rules/no-unused-vars) rule to catch. Have you tried using that?

[playground example](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC12YXJzOiBcImVycm9yXCIgKi9cblxuZnVuY3Rpb24geChhcmcpIHtcbiAgYXJnID0gMTtcbiAgcmV0dXJuIDg7XG59Iiwib3B0aW9ucyI6eyJlbnYiOnt9LCJydWxlcyI6e30sInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fSwiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0In19fQ==)

**Comment by boris-petrov** [comment]
@Rec0iL99 thanks for the response!

Maybe that rule is better suited, not sure, perhaps you'll have to decide. It still doesn't handle correctly my use case:

[playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC12YXJzOiBcImVycm9yXCIgKi9cblxuZnVuY3Rpb24geChhcmcpIHtcbiAgY29uc3QgdmFsdWUgPSBhcmc7XG4gIGFyZyA9IDg7XG4gIHJldHVybiB2YWx1ZTtcbn0iLCJvcHRpb25zIjp7ImVudiI6e30sInJ1bGVzIjp7fSwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9LCJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQifX19)

**Comment by nzakas** [comment]
The `no-unused-expressions` rule is working as intended. It intentionally omits checks for assignment expressions, function calls, and `new` calls. See https://github.com/eslint/eslint/pull/327 for background.

The `no-unused-vars` rule is also working as intended. An argument is considered used when its value is read. Because your example first reads `arg` and then writes to it, the variable is considered used. If the variable wasn't first read, `no-unused-vars` would pick it up, [like this](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC12YXJzOiBbXCJlcnJvclwiLCB7IGFyZ3M6IFwiYWxsXCJ9XSAqL1xuXG5mdW5jdGlvbiB4KGFyZykge1xuICBhcmcgPSA4O1xufSIsIm9wdGlvbnMiOnsiZW52Ijp7fSwicnVsZXMiOnt9LCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e30sImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCJ9fX0=)

**Comment by boris-petrov** [comment]
@nzakas thanks for the explanations! I'm fine with these two rules working as expected. That's why I'm proposing a rule change or a new rule being considered - to handle the case that I gave an example for. That leads to real bugs. SonarQube has a check for that so it's obviously useful. Perhaps you can reopen the issue and rename it to be a new-rule request?

**Comment by nzakas** [comment]
This behavior doesn't belong in `no-unused-expressions`. You are welcome to submit a proposal for a new rule, but you need to use the appropriate template for that. 

For others, the Sonar rule in question is here:
https://rules.sonarsource.com/javascript/RSPEC-1854/
